flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/IncrementalAggregateReduceFunction.scala

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.api.table.runtime.aggregate
+
+import org.apache.flink.api.common.functions.ReduceFunction
+import org.apache.flink.api.table.Row
+import org.apache.flink.util.Preconditions
+
+/**
+  * For Incremental intermediate aggregate Rows, merge every row into aggregate buffer.
+  *
+  * @param aggregates   The aggregate functions.
+  * @param groupKeysMapping The index mapping of group keys between intermediate aggregate Row
+  *                         and output Row.
+  */
+class IncrementalAggregateReduceFunction(
+    private val aggregates: Array[Aggregate[_]],
+    private val groupKeysMapping: Array[(Int, Int)],
+    private val intermediateRowArity: Int)extends ReduceFunction[Row] {
+
+  Preconditions.checkNotNull(aggregates)
+  Preconditions.checkNotNull(groupKeysMapping)
+  @transient var accumulatorRow:Row = _
+
+  /**
+    * For Incremental intermediate aggregate Rows, merge value1 and value2
+    * into aggregate buffer, return aggregate buffer.
+    *
+    * @param value1 The first value to combined.
+    * @param value2 The second value to combined.
+    * @return The combined value of both input values.
+    *
+    */
+  override def reduce(value1: Row, value2: Row): Row = {
+
+    if(null == accumulatorRow){
+      accumulatorRow = new Row(intermediateRowArity)
+    }
+
+    // Initiate intermediate aggregate value.
+    aggregates.foreach(_.initiate(accumulatorRow))
```

Comment:
Hi @fhueske ,  you are right, in case of sliding window, the result will be incorrect. But the `accumulatorRow` way has the same problem, because the same `accumulatorRow` object is used in multiple windows as reduce state.

Try this case 

```scala
val data = List(
    (2L, 2, "Hello"),
    (3L, 2, "Hello"),
    (4L, 2, "Hello"))

val stream = env
      .fromCollection(data)
      .assignTimestampsAndWatermarks(new TimestampWithEqualWatermark())
val table = stream.toTable(tEnv, 'long, 'int, 'string)

val windowedTable = table
      .groupBy('string)
      .window(Slide over 10.milli every 5.milli on 'rowtime as 'w)
      .select('string, 'int.count, 'w.start, 'w.end, 'w.start)
```

The expected result should be 

```
"Hello,3,1969-12-31 23:59:59.995,1970-01-01 00:00:00.005,1969-12-31 23:59:59.995",
"Hello,3,1970-01-01 00:00:00.0,1970-01-01 00:00:00.01,1970-01-01 00:00:00.0"
```

But actually it is 

```
"Hello,4,1969-12-31 23:59:59.995,1970-01-01 00:00:00.005,1969-12-31 23:59:59.995",
"Hello,4,1970-01-01 00:00:00.0,1970-01-01 00:00:00.01,1970-01-01 00:00:00.0"
```

I think it is a bug of `HeapReducingState` that the element put into (or get) State should be always a copy.  @aljoscha  what do you think about this ?